// node-opcua-factory/dist/factories_factories.js
"use strict";
/**
 * @module node-opcua-factory
 */
// tslint:disable:no-console
Object.defineProperty(exports, "__esModule", { value: true });
const node_opcua_debug_1 = require("node-opcua-debug");
const datatype_factory_1 = require("./datatype_factory");
const debugLog = node_opcua_debug_1.make_debugLog(__filename);
const doDebug = node_opcua_debug_1.checkDebugFlag(__filename);
let globalFactory;
function getStandartDataTypeFactory() {
    if (!globalFactory) {
        globalFactory = new datatype_factory_1.DataTypeFactory([]);
        globalFactory.targetNamespace = "http://opcfoundation.org/UA/";
    }
    return globalFactory;
}
exports.getStandartDataTypeFactory = getStandartDataTypeFactory;
function getStructureTypeConstructor(typeName) {
    return getStandartDataTypeFactory().getStructureTypeConstructor(typeName);
}
exports.getStructureTypeConstructor = getStructureTypeConstructor;
function hasStructuredType(typeName) {
    return getStandartDataTypeFactory().hasStructuredType(typeName);
}
exports.hasStructuredType = hasStructuredType;
function getStructuredTypeSchema(typeName) {
    return getStandartDataTypeFactory().getStructuredTypeSchema(typeName);
}
exports.getStructuredTypeSchema = getStructuredTypeSchema;
function getConstructor(binaryEncodingNodeId) {
    return getStandartDataTypeFactory().getConstructor(binaryEncodingNodeId);
}
exports.getConstructor = getConstructor;
function hasConstructor(binaryEncodingNodeId) {
    return getStandartDataTypeFactory().hasConstructor(binaryEncodingNodeId);
}
exports.hasConstructor = hasConstructor;
function constructObject(binaryEncodingNodeId) {
    return getStandartDataTypeFactory().constructObject(binaryEncodingNodeId);
}
exports.constructObject = constructObject;
function registerClassDefinition(dataTypeNodeId, className, classConstructor) {
    return getStandartDataTypeFactory().registerClassDefinition(dataTypeNodeId, className, classConstructor);
}
exports.registerClassDefinition = registerClassDefinition;
/* istanbul ignore next */
function dump() {
    getStandartDataTypeFactory().dump();
}
exports.dump = dump;
